modules/dex/src/lib.rs

```diff
@@ -851,6 +980,198 @@ impl<T: Config> Pallet<T> {
 		T::PalletId::get().into_account_truncating()
 	}
 
+	fn treasury_account() -> T::AccountId {
+		T::TreasuryPallet::get().into_account_truncating()
+	}
+
+	fn submit_triangle_swap_tx(currency_1: CurrencyId, currency_2: CurrencyId, currency_3: CurrencyId) {
+		let call = Call::<T>::triangle_swap {
+			currency_1,
+			currency_2,
+			currency_3,
+		};
+		if let Err(err) = SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into()) {
+			log::info!(
+				target: "dex-bot",
+				"offchain worker: submit unsigned swap A={:?},B={:?},C={:?}, failed: {:?}",
+				currency_1, currency_2, currency_3, err,
+			);
+		}
+	}
+
+	fn _offchain_worker(now: T::BlockNumber) -> Result<(), OffchainErr> {
+		// acquire offchain worker lock
+		let lock_expiration = Duration::from_millis(LOCK_DURATION);
+		let mut lock = StorageLock::<'_, Time>::with_deadline(OFFCHAIN_WORKER_LOCK, lock_expiration);
+		let mut guard = lock.try_lock().map_err(|_| OffchainErr::OffchainLock)?;
+		// get the max iterations config
+		let max_iterations = StorageValueRef::persistent(OFFCHAIN_WORKER_MAX_ITERATIONS)
+			.get::<u32>()
+			.unwrap_or(Some(DEFAULT_MAX_ITERATIONS))
+			.unwrap_or(DEFAULT_MAX_ITERATIONS);
+		let mut to_be_continue = StorageValueRef::persistent(OFFCHAIN_WORKER_DATA);
+
+		// update `TradingPairNodes` every `TradingKeysUpdateFrequency` interval.
+		if now % T::TradingKeysUpdateFrequency::get() == Zero::zero() {
+			let mut trading_pair_values_map: BTreeMap<CurrencyId, Vec<CurrencyId>> = BTreeMap::new();
+			TradingPairStatuses::<T>::iter()
+				.filter(|(_, status)| status.enabled())
+				.for_each(|(pair, _)| {
+					trading_pair_values_map
+						.entry(pair.first())
+						.or_insert_with(Vec::<CurrencyId>::new)
+						.push(pair.second());
+				});
+			for (currency_id, trading_tokens) in trading_pair_values_map {
+				let _ = TradingPairNodes::<T>::try_mutate(currency_id, |maybe_trading_tokens| -> DispatchResult {
```

Comment:
I don't think you can mutate storages in offchain worker? at least the storages won't be saved. you need to use offchain worker db for that

Reply:
`TradingPairNodes` type is CurrencyId->[CurrencyId], using offchain db I think it's too complicate. Current I'm moving this storage mutate part to `on_idle`.

Reply:
remove `TradingPairNodes`, and calculation without on chain storage.